packages/controllers/src/permissions/PermissionController.ts

```diff
@@ -0,0 +1,1970 @@
+   * @returns Whether the specified request exists.
+   */
+  private hasApprovalRequest(options: { id: string }): boolean {
+    return this.messagingSystem.call(
+      'ApprovalController:hasRequest',
+      // Typecast: For some reason, the type here expects all of the possible
+      // HasApprovalRequest options to be specified, when they're actually all
+      // optional. Passing just the id is definitely valid, so we just cast it.
+      options as any,
+    );
+  }
+
+  /**
+   * Rejects the permissions request with the specified id, with the specified
+   * error as the reason. This method is effectively a wrapper around a
+   * messenger call for the `ApprovalController:rejectRequest` action.
+   *
+   * @see {@link PermissionController.acceptPermissionsRequest} and
+   * {@link PermissionController.rejectPermissionsRequest} for usage.
+   * @param id - The id of the request to reject.
+   * @param error - The error associated with the rejection.
+   */
+  private _rejectPermissionsRequest(id: string, error: Error): void {
+    return this.messagingSystem.call(
+      'ApprovalController:rejectRequest',
+      id,
+      error,
+    );
+  }
+
+  /**
+   * Executes a restricted method as the subject with the the given origin.
+   * The specified params, if any, will be passed to the method implementation.
+   *
+   * **ATTN:** Great caution should be exercised in the use of this method.
+   * Methods that cause side effects or affect application state should
+   * be avoided.
+   *
+   * This method will first attempt to retrieve the requested restrictded method
+   * implementation, throwing if it does not exist. The method will then be
+   * invoked as though the subject with the specified origin had invoked it with
+   * the specified parameters. This means that any existing caveats will be
+   * applied to the restricted method, and this method will throw if the
+   * restricted method or its caveat decorators throw.
+   *
+   * In addition, this method will throw if the subject does not have a
+   * permission for the specified restricted method.
+   *
+   * @param origin - The origin of the subject to execute the method on behalf
+   * of.
+   * @param methodName - The name of the method to execute.
+   * @param params - The parameters to pass to the method implementation.
+   * @returns The result of the executed method.
+   */
+  async executeRestrictedMethod(
+    origin: OriginString,
+    methodName: ExtractPermission<
+      ControllerPermissionSpecification,
+      ControllerCaveatSpecification
+    >['parentCapability'],
+    params?: RestrictedMethodParameters,
+  ): Promise<Json> {
+    // Throws if the method does not exist
+    const methodImplementation = this._getRestrictedMethod(methodName, origin);
+
+    const result = await this._executeRestrictedMethod(
+      methodImplementation,
+      { origin },
+      methodName,
+      params,
+    );
+
+    if (result === undefined) {
+      throw new Error(
```

Comment:
Do all RPC methods return some result? I would have assumed that was allowed - I'm not really sure either way though.

Reply:
A JSON-RPC response object must contain either the `result` or `error` property, and since the `result` can't be `undefined`, I think it's fair to disallow restricted methods returning `undefined` also. `void` methods will have to return `null`.

Reply:
Ah right! Makes sense, thanks